### PR TITLE
Add `PlutusType` derivation support for data encoded types

### LINKED ISSUE #187: `instance PIsDataRepr a => PlutusType (PIsDataReprInstances a)`
The missing piece is `pcon`. We need to use generics to implement it.

### PR DESCRIPTION
**Pull request by TotallyNotChase**
Subset of #242 

Resolves #187

### CODE REVIEW

File: Plutarch/DataRepr.hs [hunk 1]
```diff
@@ -22,5 +24,27 @@ module Plutarch.DataRepr (
...
+You can specify the label to associate with the field using type applications-
+
+@
+
+foo :: Term s (PDataRecord '[ "fooField" ':= PByteString ])
+foo = pdcons @"fooField" # pdata (phexByteStr "ab") # pdnil
+
+@
+-}
+pdcons :: forall label a l s. Term s (PAsData a :--> I.PDataRecord l :--> I.PDataRecord ((label 'I.:= a) ': l))
```
Comment (TotallyNotChase) [inline Plutarch/DataRepr.hs:45]: I wasn't sure about the type variable ordering here. Should label come before or after a - which one would users be more likely to use type application for?

**Comment by L-as** [comment]
`pmatch` should be optimally efficient AFAICT, because you already know whether it's a cons or not.

**Comment by TotallyNotChase** [comment]
> `pmatch` should be optimally efficient AFAICT, because you already know whether it's a cons or not.

That's fine - we don't have an issue with efficiency here but I'm more concerned about what this would achieve other than add in more concepts to `PDataRecord` that users *shouldn't* need. It gives the impression that `PDataRecord` is just another general purpose plutarch type that can be used everywhere. But I thought it was more domain specific, particularly for encoding product types and accesssing them. We *have* `PBuiltinList` and its plutus type instance already after all.

There'd be 2 ways to match on data records. There'd probably also be 2 ways to construct them because `pcon $ PDCons x xs`, or `pcon $ PDCons @"label" x xs` isn't as convenient as `pdcons # x # xs` and `pdcons @"label" # x # xs`.

FWIW, I did implement `PlutusType` locally just to see what it looks like. I just don't see a point in doing it, and I'm concerned about the implications such an instance gives.

**Comment by L-as** [comment]
`PDataRecord` is a general purpose type though.

**Comment by TotallyNotChase** [comment]
Alright, I added a `PlutusType` instance for `PDataRecord`

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,28 @@
...
+
+  All existing ledger api types now have `PlutusType` instances - not just `PMatch`.
+
+  Started by: [#250](https://github.com/Plutonomicon/plutarch/pull/250)
+
+- Add `PDataRecord` construction utilities, necessary for full usage of data encoded `PlutusType` instances.
+
+  In particular, you can build `PDataRecord`s with `pdcon` and `pdnil` - refer to the guide for more info.
+
+  `pdcon` and `pdnil` are also exported from `Plutarch.Prelude`.
```
Comment (L-as) [inline CHANGELOG.md:21]: `pdcon`?

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,3 +1,3 @@
 {
     "haskell.formattingProvider": "fourmolu"
-}
+}
```
Comment (L-as) [inline .vscode/settings.json:3]: Why did you remove the newline here?

File: Plutarch/DataRepr/Internal.hs [hunk 4]
```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
...
  `PBuiltinList PData`.
 -}
-data PDataRecord (as :: [PLabeledType]) (s :: S)
+data PDataRecord (as :: [PLabeledType]) (s :: S) where
+  PDCons ::
+    forall name x xs s.
+    Term s (PAsData x) ->
+    (Term s (PDataRecord xs)) ->
+    PDataRecord ((name ':= x) ': xs) s
+  PDNil :: PDataRecord '[] s
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:112]: Looks great

File: Plutarch/DataRepr/Internal.hs [hunk 5]
```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
...
+
+instance PlutusType (PDataRecord ((name ':= x) ': xs)) where
+  type PInner (PDataRecord ((name ':= x) ': xs)) _ = PBuiltinList PData
+  pcon' (PDCons x xs) = punsafeCoerce result
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:116]: There's no need to use `punsafeCoerce` here. Just do `pto`.

File: Plutarch/DataRepr/Internal.hs [hunk 6]
```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
...
+      result :: Term _ (PDataRecord ((name ':= x) ': xs))
+      result = pdcons # x # xs
+  pmatch' x f = pmatch x $ \case
+    PCons x' xs' ->
+      let x :: Term _ (PAsData x)
+          x = punsafeCoerce x'
+          xs :: Term _ (PDataRecord xs)
+          xs = punsafeCoerce xs'
+       in f $ PDCons x xs
+    PNil -> error "PDataRecord:pmatch':PNil:absurd"
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:127]: This will cause any code that uses `pmatch` on `PDataRecord` to err at compile-time unconditionally.

File: Plutarch/DataRepr/Internal.hs [hunk 7]
```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
...
+  PDNil :: PDataRecord '[] s
+
+instance PlutusType (PDataRecord ((name ':= x) ': xs)) where
+  type PInner (PDataRecord ((name ':= x) ': xs)) _ = PBuiltinList PData
+  pcon' (PDCons x xs) = punsafeCoerce result
+    where
+      result :: Term _ (PDataRecord ((name ':= x) ': xs))
+      result = pdcons # x # xs
+  pmatch' x f = pmatch x $ \case
+    PCons x' xs' ->
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:121]: We already know the list is non-empty. The reason for this is that on line 114, we declare the instance for only non-empty data records. You can just use `phead`.

File: Plutarch/DataRepr/Internal.hs [hunk 8]
```diff
@@ -65,17 +91,63 @@ import Plutarch.Builtin (
...
+      let x :: Term _ (PAsData x)
+          x = punsafeCoerce x'
+          xs :: Term _ (PDataRecord xs)
+          xs = punsafeCoerce xs'
+       in f $ PDCons x xs
+    PNil -> error "PDataRecord:pmatch':PNil:absurd"
+
+instance PlutusType (PDataRecord '[]) where
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:129]: Great job BTW splitting up the instance into two.

File: Plutarch/DataRepr/Internal.hs [hunk 9]
```diff
@@ -203,10 +275,29 @@ pmatchDataSum d handlers =
 -}
 newtype PIsDataReprInstances (a :: PType) (s :: S) = PIsDataReprInstances (a s)
 
+-- Q: Why are these superclass constraints here?
 class (PMatch a, PIsData a) => PIsDataRepr (a :: PType) where
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:279]: `PMatch a` should be replaced with `PlutusType a`. This is to make sure people don't forget to derive all necessary instances. `PIsData a` helps avoid accidentally unsound code.

File: examples/Examples/ConstrData.hs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    (PlutusType, PIsData, PDataFields)
+    via (PIsDataReprInstances (Triplet a))
+
+data PVehicle (s :: S)
+  = PFourWheeler (Term s (PDataRecord '["_0" ':= PInteger, "_1" ':= PInteger, "_2" ':= PInteger, "_3" ':= PInteger]))
+  | PTwoWheeler (Term s (PDataRecord '["_0" ':= PInteger, "_1" ':= PInteger]))
+  | PImmovableBox (Term s (PDataRecord '[]))
+  deriving stock (GHC.Generic)
+  deriving anyclass (Generic)
+  deriving anyclass (PIsDataRepr)
```
Comment (L-as) [inline examples/Examples/ConstrData.hs:53]: Why have separate lines for the anyclass derives?

**Review by L-as — CHANGES_REQUESTED** [review]
Looks mostly great!

File: Plutarch/DataRepr/Internal.hs (hunk 6 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs:127]: oh I meant to write `perror` - whoops.

File: Plutarch/DataRepr/Internal.hs (hunk 7 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs:121]: Yes, but we need the head and tail. `ChooseList` (what `pmatch` will expand to here) is going to more efficient than head + tail. At the cost of also having to handle the absurd nil case.

File: examples/Examples/ConstrData.hs (hunk 10 above)
Comment (TotallyNotChase) [reply examples/Examples/ConstrData.hs:53]: Yeah we've kinda been following this style throughout the entire repo.

File: Plutarch/DataRepr/Internal.hs (hunk 9 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs]: Sounds good. Updated!

File: Plutarch/DataRepr/Internal.hs (hunk 5 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs]: Aha! Good idea! Updated.

File: .vscode/settings.json (hunk 3 above)
Comment (TotallyNotChase) [reply .vscode/settings.json]: I have no idea - it was probably done by vscode and I didn't notice.

This is a bit of an issue since I have my own local settings in this file and I have to remove them everytime I check out branches.... oh well.

File: CHANGELOG.md (hunk 2 above)
Comment (TotallyNotChase) [reply CHANGELOG.md]: Good catch! Fixed.

File: .vscode/settings.json (hunk 3 above)
Comment (L-as) [reply .vscode/settings.json]: If it's done by VSCode, then don't care about it. Just VSCode being stupid I suppose.

Comment (L-as) [reply .vscode/settings.json]: What are your own local settings? Would it be best if we remove the entire file entirely again? You might want to discuss this with @srid , I don't use VSCode.

File: Plutarch/DataRepr/Internal.hs (hunk 7 above)
Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:121]: I'm not entirely sure what you mean. `pmatch` also just gets the head and tail separately, it only uses `ChooseList` to figure out whether it's a head or tail IIRC. Look at your own PR! https://github.com/Plutonomicon/plutarch/pull/217/files

File: examples/Examples/ConstrData.hs (hunk 10 above)
Comment (L-as) [reply examples/Examples/ConstrData.hs:53]: It's been on accident, it's not a good style.

Comment (TotallyNotChase) [reply examples/Examples/ConstrData.hs:53]: We probably want to fix that as a cleanup task - do you want it done here in this example for the time being?

File: Plutarch/DataRepr/Internal.hs (hunk 7 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs]: Oh lol - I was under the impression choose list takes in lambdas. Never mind! Updated.

File: examples/Examples/ConstrData.hs (hunk 10 above)
Comment (L-as) [reply examples/Examples/ConstrData.hs:53]: Yeah let's make sure we don't add more bad code.

File: Plutarch/DataRepr/Internal.hs [hunk 11]
```diff
@@ -276,8 +275,7 @@ pmatchDataSum d handlers =
 -}
 newtype PIsDataReprInstances (a :: PType) (s :: S) = PIsDataReprInstances (a s)
 
--- Q: Why are these superclass constraints here?
-class (PMatch a, PIsData a) => PIsDataRepr (a :: PType) where
+class (PlutusType a, PIsData a) => PIsDataRepr (a :: PType) where
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:275]: I just realised this is a breaking change... sorry. Can we revert this commit? lol

Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs:275]: Hmm, breaking because users may not be implementing `PlutusType` yet? Don't we *want* them to do just that though, from now on?

But yes, we can revert it if you want.

Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:275]: Existing code would fail to compile because `PlutusType` isn't derived.

Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs]: Reverted - made a note to change this in the future.

File: examples/Examples/ConstrData.hs (hunk 10 above)
Comment (TotallyNotChase) [reply examples/Examples/ConstrData.hs]: done!

**Review by L-as — APPROVED** [review]
LGTM
